File: app/src/main/java/com/worldventures/dreamtrips/social/ui/tripsimages/presenter/MemberImagesPresenter.java
package com.worldventures.dreamtrips.social.ui.tripsimages.presenter;

import com.innahema.collections.query.queriables.Queryable;
import com.worldventures.dreamtrips.R;
import com.worldventures.dreamtrips.social.ui.background_uploading.model.CompoundOperationState;
import com.worldventures.dreamtrips.social.ui.tripsimages.model.BaseMediaEntity;
import com.worldventures.dreamtrips.social.ui.tripsimages.service.command.CheckVideoProcessingStatusCommand;
import com.worldventures.dreamtrips.social.ui.tripsimages.service.command.MemberImagesAddedCommand;
import com.worldventures.dreamtrips.social.ui.tripsimages.service.delegate.MediaRefresher;
import com.worldventures.dreamtrips.social.ui.tripsimages.view.args.TripImagesArgs;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import rx.Observable;
import timber.log.Timber;

public class MemberImagesPresenter extends TripImagesPresenter {

   @Inject MediaRefresher memberImagesRefresher;

   public MemberImagesPresenter(TripImagesArgs tripImagesArgs) {
      super(tripImagesArgs);
   }

   @Override
   public void onResume() {
      super.onResume();
      subscribeToRefresher();
   }

   @Override
   public void onPause() {
      super.onPause();
      stopRefreshingMemberImages();
   }

   void subscribeToRefresher() {
      tripImagesInteractor.baseTripImagesCommandActionPipe().observeSuccess()
            .compose(bindUntilPauseIoToMainComposer())
            .subscribe(getMembersPhotosCommand -> startRefreshingMemberImages());
      memberImagesRefresher.getNewPhotosObservable()
            .compose(bindUntilPauseIoToMainComposer())
            .subscribe(this::processNewPhotosFromRefresher,
                  throwable -> Timber.w(throwable, "Could not refresh trip images"));
      if (!loading) startRefreshingMemberImages();
   }

   public void onShowNewImagesClick() {
      List<BaseMediaEntity> lastPhotos = new ArrayList<>(memberImagesRefresher.getLastPhotos());
      if (lastPhotos.size() < tripImagesArgs.getPageSize()) {
         tripImagesInteractor.memberImagesAddedCommandPipe()
               .send(new MemberImagesAddedCommand(tripImagesArgs, lastPhotos));
         currentItems.addAll(0, lastPhotos);
         tripImagesInteractor.checkVideoProcessingStatusPipe()
               .send(new CheckVideoProcessingStatusCommand(currentItems));
         updateItemsInView();
         restartImageRefreshing();
      } else {
         reload();
         stopRefreshingMemberImages();
      }
      view.hideNewImagesButton();
   }

   private void startRefreshingMemberImages() {
      if (memberImagesAreRefreshing) return;
      memberImagesAreRefreshing = true;
      memberImagesRefresher.startRefreshing(Observable.defer(() -> {
         if (currentItems.isEmpty()) return Observable.empty();
         return Observable.just(currentItems.get(0));
      }), tripImagesArgs);
   }

   private void stopRefreshingMemberImages() {
      memberImagesRefresher.stopRefreshing();
      memberImagesAreRefreshing = false;
   }

   private void restartImageRefreshing() {
      stopRefreshingMemberImages();
      startRefreshingMemberImages();
   }

   private void processNewPhotosFromRefresher(List<BaseMediaEntity> newPhotos) {
      boolean uploading = Queryable.from(compoundOperationModels)
            .filter(element -> element.state() == CompoundOperationState.STARTED
                  || element.state() == CompoundOperationState.FINISHED)
            .count() > 0;
      if (loading || uploading) return;
      int newPhotosCount = Queryable.from(newPhotos).filter(photo -> !currentItems.contains(photo)).count();
      if (newPhotosCount > 0) {
         String newImagesString;
         String photosCountString = String.valueOf(newPhotosCount);
         if (newPhotosCount >= tripImagesArgs.getPageSize()) {
            newImagesString = context.getString(R.string.member_images_new_items_plus, photosCountString);
         } else if (newPhotosCount == 1) {
            newImagesString = context.getString(R.string.member_images_new_item, photosCountString);
         } else {
            newImagesString = context.getString(R.string.member_images_new_items, photosCountString);
         }
         view.showNewImagesButton(newImagesString);
      }
   }
}

File: app/src/main/java/com/worldventures/dreamtrips/social/ui/tripsimages/presenter/MemberImagesPresenter.kt
package com.worldventures.dreamtrips.social.ui.tripsimages.presenter

import com.worldventures.dreamtrips.R
import com.worldventures.dreamtrips.social.ui.background_uploading.model.CompoundOperationState
import com.worldventures.dreamtrips.social.ui.tripsimages.model.BaseMediaEntity
import com.worldventures.dreamtrips.social.ui.tripsimages.service.command.BaseMediaCommand
import com.worldventures.dreamtrips.social.ui.tripsimages.service.command.CheckVideoProcessingStatusCommand
import com.worldventures.dreamtrips.social.ui.tripsimages.service.command.MemberImagesAddedCommand
import com.worldventures.dreamtrips.social.ui.tripsimages.service.delegate.MediaRefresher
import com.worldventures.dreamtrips.social.ui.tripsimages.view.args.TripImagesArgs
import rx.Observable
import timber.log.Timber
import java.util.ArrayList
import javax.inject.Inject

class MemberImagesPresenter(tripImagesArgs: TripImagesArgs) : TripImagesPresenter(tripImagesArgs) {

   @Inject lateinit var memberImagesRefresher: MediaRefresher

   override fun onResume() {
      super.onResume()
      subscribeToRefresher()
   }

   override fun onPause() {
      super.onPause()
      stopRefreshingMemberImages()
   }

   private fun subscribeToRefresher() {
      tripImagesInteractor.baseTripImagesPipe.observeSuccess()
            .compose<BaseMediaCommand>(bindUntilPauseIoToMainComposer())
            .subscribe { startRefreshingMemberImages() }
      memberImagesRefresher.newPhotosObservable
            .compose(bindUntilPauseIoToMainComposer())
            .subscribe(this::processNewPhotosFromRefresher,
                  { Timber.w(it, "Could not refresh trip images") })
      if (!loading) {
         startRefreshingMemberImages()
      }
   }

   fun onShowNewImagesClick() {
      val lastPhotos = ArrayList(memberImagesRefresher.lastPhotos)
      if (lastPhotos.size < tripImagesArgs.pageSize) {
         tripImagesInteractor.memberImagesAddedCommandPipe.send(MemberImagesAddedCommand(tripImagesArgs, lastPhotos))
         currentItems.addAll(0, lastPhotos)
         tripImagesInteractor.checkVideoProcessingStatusPipe.send(CheckVideoProcessingStatusCommand(currentItems))
         updateItemsInView()
         restartImageRefreshing()
      } else {
         reload()
         stopRefreshingMemberImages()
      }
      view.hideNewImagesButton()
   }

   private fun startRefreshingMemberImages() {
      if (memberImagesAreRefreshing) {
         return
      }
      memberImagesAreRefreshing = true
      memberImagesRefresher.startRefreshing(Observable.defer {
         if (currentItems.isEmpty()) {
            return@defer Observable.empty<BaseMediaEntity<*>>()
         }
         return@defer Observable.just(currentItems[0])
      }, tripImagesArgs)
   }

   private fun stopRefreshingMemberImages() {
      memberImagesRefresher.stopRefreshing()
      memberImagesAreRefreshing = false
   }

   private fun restartImageRefreshing() {
      stopRefreshingMemberImages()
      startRefreshingMemberImages()
   }

   private fun processNewPhotosFromRefresher(newPhotos: List<BaseMediaEntity<*>>) {
      val uploading = compoundOperationModels.count {
         val state = it.state()
         state == CompoundOperationState.STARTED || state == CompoundOperationState.FINISHED } > 0
      if (loading || uploading) {
         return
      }
      val newPhotosCount = newPhotos.count { !currentItems.contains(it) }
      if (newPhotosCount > 0) {
         view.showNewImagesButton(formatNewImagesString(newPhotosCount))
      }
   }

   private fun formatNewImagesString(newPhotosCount: Int): String {
      val photosCountString = newPhotosCount.toString()
      return when {
         newPhotosCount >= tripImagesArgs.pageSize -> context
               .getString(R.string.member_images_new_items_plus, photosCountString)
         newPhotosCount == 1 -> context.getString(R.string.member_images_new_item, photosCountString)
         else -> context.getString(R.string.member_images_new_items, photosCountString)
      }
   }
}
